axelor-contract/src/main/java/com/axelor/apps/contract/batch/BatchContractRevaluate.java

```diff
@@ -0,0 +1,218 @@
+  protected ContractRevaluationService contractRevaluationService;
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  @Inject
+  public BatchContractRevaluate(
+      ContractService contractService,
+      ContractRepository contractRepository,
+      ContractVersionService contractVersionService,
+      ContractRevaluationService contractRevaluationService) {
+    this.contractService = contractService;
+    this.contractRepository = contractRepository;
+    this.contractVersionService = contractVersionService;
+    this.contractRevaluationService = contractRevaluationService;
+  }
+
+  protected void process() {
+    revaluateContracts();
+  }
+
+  protected void revaluateContracts() {
+    List<Long> idsOk = new ArrayList<>();
+    List<Long> idsFail = new ArrayList<>();
+    List<Long> idsReevaluated = new ArrayList<>();
+    idsOk.add(0L);
+    idsFail.add(0L);
+    Query<Contract> query =
+        contractRepository
+            .all()
+            .filter(
+                "self.isToRevaluate = true AND self.currentContractVersion.statusSelect = :onGoingStatus AND self.id NOT IN (:idsOk) AND self.id NOT IN (:idsFail)");
+    List<Contract> contractList;
+    while (!(contractList =
+            query
+                .bind("onGoingStatus", ContractVersionRepository.ONGOING_VERSION)
+                .bind("idsOk", idsOk)
+                .bind("idsFail", idsFail)
+                .fetch(FETCH_LIMIT))
+        .isEmpty()) {
+      Map<String, List<Contract>> ids = revaluateContracts(contractList);
+      idsOk.addAll(ids.get("OK").stream().map(Contract::getId).collect(Collectors.toList()));
+      idsFail.addAll(ids.get("FAIL").stream().map(Contract::getId).collect(Collectors.toList()));
+      idsReevaluated.addAll(
+          ids.get("REEVALUATED").stream().map(Contract::getId).collect(Collectors.toList()));
+      JPA.clear();
+    }
+    LOG.debug("{} Reevaluated contracts : {}", idsReevaluated.size(), idsReevaluated);
+  }
+
+  protected Map<String, List<Contract>> revaluateContracts(List<Contract> contractList) {
+    List<Contract> idsFail = new ArrayList<>();
+    List<Contract> idsOk = new ArrayList<>();
+    List<Contract> idsReevaluated = new ArrayList<>();
+    for (Contract contract : contractList) {
+      processContract(idsFail, idsOk, idsReevaluated, contract);
+    }
+    return Stream.of(
+            new AbstractMap.SimpleEntry<>("OK", idsOk),
+            new AbstractMap.SimpleEntry<>("FAIL", idsFail),
+            new AbstractMap.SimpleEntry<>("REEVALUATED", idsReevaluated))
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+  }
+
+  @Transactional(rollbackOn = Exception.class)
+  protected void processContract(
+      List<Contract> idsFail,
+      List<Contract> idsOk,
+      List<Contract> idsReevaluated,
+      Contract contract) {
+    try {
+      contract = contractRepository.find(contract.getId());
```

Comment:
I don't think you need to find again the contract here, since it is already returned by the query.